Add support for source on update self command

nkolev92 opened the pull request:
## Bug

Fixes: https://github.com/NuGet/Home/issues/1783
Regression: No  
* Last working version:   
* How are we preventing it in future:   

## Fix

Details: 
Add support for passing exactly one source when calling nuget update self.
note that the source neeeds to be explcitly passed and won't use the default config sources cause that'd be a breaking change. 

Lastly we only support 1 exact source. 

## Testing/Validation

Tests Added: Yes  
Reason for not adding tests:  
Validation:  Manual & tests. Also verified it works with password protected sources with creds in config.

nkolev92 committed:
Update self can accept an argument

nkolev92 committed:
fix test

donnie-msft commented on src/NuGet.Clients/NuGet.CommandLine/Commands/UpdateCommand.cs:
```diff
@@ -131,6 +131,26 @@ public override async Task ExecuteCommandAsync()
+                    targetSource = Common.PackageSourceProviderExtensions.ResolveSource(availableSources, Source.Single());
+                    break;
+                default:
+                    throw new CommandException(NuGetResources.Errror_UpdateSelf_Source);
```
typo: Error

nkolev92 replied:
fixed.

donnie-msft commented on src/NuGet.Clients/NuGet.CommandLine/NuGetResources.resx:
```diff
@@ -5417,4 +5417,7 @@ Oluşturma sırasında NuGet'in paketleri indirmesini önlemek için, Visual Stu
     <value>Invalid value given for 'DependencyVersion': "{0}".</value>
     <comment>Please don't localize "DependencyVersion".</comment>
   </data>
+  <data name="Errror_UpdateSelf_Source" xml:space="preserve">
```
typo: Error

donnie-msft commented on src/NuGet.Clients/NuGet.CommandLine/NuGet.CommandLine.csproj:
```diff
@@ -124,7 +124,7 @@
   <!-- 	Since we are moving some code and strings from NuGet.CommandLine to NuGet.Commands, we opted to go through normal localization process
         (build .resources.dll) and then add them to the ILMerged nuget.exe -->
   <!-- This will also be called from CI build, after assemblies are localized, since our test infra takes nuget.exe before Localization  -->
-  <Target Name="ILMergeNuGetExe" AfterTargets="Build" Condition="'$(BuildingInsideVisualStudio)' != 'true' and '$(SkipILMergeOfNuGetExe)' != 'true'" >
```
Is this unrelated?

nkolev92 replied:
I'll revert this. You can't rerun a test in VS that requires the IL merged version of the exe.

donnie-msft commented on src/NuGet.Clients/NuGet.CommandLine/GlobalSuppressions.cs:
```diff
@@ -64,7 +64,6 @@
 [assembly: SuppressMessage("Build", "CA1307:The behavior of 'string.Equals(string, string)' could vary based on the current user's locale settings. Replace this call in 'NuGet.CommandLine.RestoreCommand.IsPackagesConfig(string)' with a call to 'string.Equals(string, string, System.StringComparison)'.", Justification = "<Pending>", Scope = "member", Target = "~M:NuGet.CommandLine.RestoreCommand.Is ...
 [assembly: SuppressMessage("Build", "CA1031:Modify 'GetNuGetVersion' to catch a more specific allowed exception type, or rethrow the exception.", Justification = "<Pending>", Scope = "member", Target = "~M:NuGet.CommandLine.SelfUpdater.GetNuGetVersion(System.Reflection.Assembly)~NuGet.Versioning.NuGetVersion")]
 [assembly: SuppressMessage("Build", "CA1031:Modify 'UpdateSelfFromVersionAsync' to catch a more specific allowed exception type, or rethrow the exception.", Justification = "<Pending>", Scope = "member", Target = "~M:NuGet.CommandLine.SelfUpdater.UpdateSelfFromVersionAsync(System.String,System.Boolean,NuGet.Versioning.NuGetVersion,System.String,System.Threading.CancellationToken)~System.Threading ...
-[assembly: SuppressMessage("Build", "CA2000:Call System.IDisposable.Dispose on object created by 'new SourceCacheContext()' before all references to it are out of scope.", Justification = "<Pending>", Scope = "member", Target = "~M:NuGet.CommandLine.SelfUpdater.UpdateSelfFromVersionAsync(System.String,System.Boolean,NuGet.Versioning.NuGetVersion,System.String,System.Threading.CancellationToken)~S ...
```
Is this unrelated?

nkolev92 replied:
I just fixed the warning, So the suppression is not necessary.

donnie-msft commented on src/NuGet.Clients/NuGet.CommandLine/Common/SelfUpdater.cs:
```diff
@@ -61,78 +60,80 @@ internal string AssemblyLocation
+                            Directory.Delete(tempDir, recursive: true);
+                        }
+                        catch
+                        {
```
A message? "hey, go look in tempDir, because we left some stuff there"

nkolev92 replied:
There's no benefit from it.
The temp folders will get deleted eventually. 

This is a just in case thing. 
Note that this is not new code, most of these are whitespace changes because of the extra using.

donnie-msft commented on test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetUpdateCommandTests.cs:
```diff
@@ -1326,7 +1324,7 @@ public void UpdateCommand_FromProjectConfig(string configFileName)
 
                 // Assert
-                Assert.True(_successCode == r.Item1, r.Item2 + " " + r.Item3);
+                Assert.True(0 == r.Item1, r.Item2 + " " + r.Item3);
```
```suggestion
                Assert.True(r.Success , r.Output + " " + r.Errors);
```

donnie-msft commented on src/NuGet.Clients/NuGet.CommandLine/Commands/UpdateCommand.cs:
```diff
@@ -131,6 +131,26 @@ public override async Task ExecuteCommandAsync()
+                    break;
+                case 1:
+                    // Use the package source from the load config to preload any creds that might be needed for authentication.
+                    var availableSources = SourceProvider.LoadPackageSources().Where(source => source.IsEnabled).ToList();
```
I don't like that this duplicates code from the Extension... and I'm not convinced with an initial look that we need to build the entire List in this layer. 
If duplicating, why not just duplicate the [ResolveSource logic](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Commands/Utility/PackageSourceProviderExtensions.cs#L17)?

Something like this avoids loading N sources up front?...
```suggestion
                    var availableSources = SourceProvider.LoadPackageSources().FirstOrDefault(f =>f.Source.IsEnabled && (f.Source.Equals(source, StringComparison.OrdinalIgnoreCase) || f.Name.Equals(source, StringComparison.OrdinalIgnoreCase))).ToList();
```

nkolev92 replied:
I removed the ToList so now it's happening lazily. 

Now I don't think we need to duplicate any code.

donnie-msft reviewed (commented):
Neat PR.. what resurrected this issue?

nkolev92 committed:
Update test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetUpdateCommandTests.cs

Co-Authored-By: donnie-msft <49205731+donnie-msft@users.noreply.github.com>

nkolev92 committed:
address feedback

nkolev92 commented:
@donnie-msft Addressed your feedback. 
There's 2 open points, PTAL. 

As far as what resurrected the issue, @aortiz-msft just pinged me on it last Friday and I just needed to write some code :D

nkolev92 committed:
Merge branch 'dev-nkolev92-addSupportForSourceOnUpdateSelfCommand' of https://github.com/NuGet/NuGet.Client into dev-nkolev92-addSupportForSourceOnUpdateSelfCommand

nkolev92 committed:
work

donnie-msft commented on src/NuGet.Clients/NuGet.CommandLine/Commands/UpdateCommand.cs:
```diff
@@ -73,8 +74,7 @@ public override async Task ExecuteCommandAsync()
             if (Self)
             {
-                var selfUpdater = new SelfUpdater(Console);
-                await selfUpdater.UpdateSelfAsync(Prerelease, NuGetConstants.V3FeedUrl);
```
Here's a comment
